### PR TITLE
Also report number of virtual services known to pilot.

### LINKED ISSUE #14932: Pilot exposes metric pilot_virt_services but never sets it
Pilot has a metric `pilot_virt_services`, but it is never actually set. This is also in the dashboard so it is misleading.

I am fine with not having it if it isn't feasible due to some code changes or something, but we should remove it from the dashboard if that is the case.

### PR DESCRIPTION
**Pull request by c0d1ngm0nk3y**
Only count virtual services for ISTIO_MESH_GATEWAY.

Fixes #14932

### CODE REVIEW

**Comment by c0d1ngm0nk3y** [comment]
The test `circleci lint` fails with

```
line 40: 23106 Killed
```

local `go lint` and `goimports` are fine. `make lint` shows a bunch of unrelated errors.

**Comment by howardjohn** [comment]
Our linter is starting to OOM on circleci (16gb....) so we are moving it to prow - the switch should be done today. I'll retry it and if not in a couple hours it will be non-required and should get merged even if its failing I think

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 1]
```diff
@@ -823,6 +823,12 @@ func (s *DiscoveryServer) AdsPushAll(version string, push *model.PushContext,
 		version, len(push.Services(nil)), adsClientCount())
 	monServices.Set(float64(len(push.Services(nil))))
 
+	onlyMesh := make(map[string]bool)
+	onlyMesh[model.IstioMeshGateway] = true
+	numberOfKnownMeshVirtualServices := len(push.VirtualServices(nil, onlyMesh))
+	adsLog.Debugf("Known virtual services for mesh gateway:%d", numberOfKnownMeshVirtualServices)
+	monVServices.Set(float64(numberOfKnownMeshVirtualServices))
```
Comment (howardjohn) [inline pilot/pkg/proxy/envoy/v2/ads.go]: I think this is still a bit misleading to scope it just to the gateways. Some people don't even use gateways for example. Maybe we can just set it in `initVirtualServices`? At that point we definitely have the correct number of virtual services

Comment (c0d1ngm0nk3y) [reply pilot/pkg/proxy/envoy/v2/ads.go]: That sounds reasonable, I will have a look. Thanks for the hint.

**Comment by kramerul** [comment]
I declare my consent.

**Comment by howardjohn** [comment]
@c0d1ngm0nk3y the racetest failure looks like it is a real one related to this PR

**Comment by kramerul** [comment]
Data races:

```
==================
WARNING: DATA RACE
Read at 0x000002a3ffc0 by goroutine 7:
  istio.io/istio/vendor/go.opencensus.io/stats/view.(*collector).addSample()
      /go/src/istio.io/istio/vendor/go.opencensus.io/stats/view/collector.go:36 +0x91
  istio.io/istio/vendor/go.opencensus.io/stats/view.(*viewInternal).addSample()
      /go/src/istio.io/istio/vendor/go.opencensus.io/stats/view/view.go:161 +0x15d
  istio.io/istio/vendor/go.opencensus.io/stats/view.(*recordReq).handleCommand()
      /go/src/istio.io/istio/vendor/go.opencensus.io/stats/view/worker_commands.go:166 +0x351
  istio.io/istio/vendor/go.opencensus.io/stats/view.(*worker).start()
      /go/src/istio.io/istio/vendor/go.opencensus.io/stats/view/worker.go:156 +0x2e7

Previous write at 0x000002a3ffc0 by goroutine 10:
  istio.io/istio/vendor/github.com/lukechampine/freeze.Slice()
      /go/src/istio.io/istio/vendor/github.com/lukechampine/freeze/freeze.go:143 +0x194
  istio.io/istio/pilot/pkg/networking/core/v1alpha3/fakes.(*frozenIstioConfigStore).List()
      /go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/fakes/frozen_istio_config_store.go:32 +0xe9
  istio.io/istio/pilot/pkg/model.(*PushContext).initSidecarScopes()
      /go/src/istio.io/istio/pilot/pkg/model/push_context.go:808 +0xc3
  istio.io/istio/pilot/pkg/model.(*PushContext).InitContext()
      /go/src/istio.io/istio/pilot/pkg/model/push_context.go:594 +0x27c
  istio.io/istio/pilot/pkg/networking/core/v1alpha3/loadbalancer.buildEnvForClustersWithDistribute()
      /go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer_test.go:195 +0x855
  istio.io/istio/pilot/pkg/networking/core/v1alpha3/loadbalancer.TestApplyLocalitySetting.func1.1()
      /go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer_test.go:63 +0x9b
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:865 +0x163

Goroutine 7 (running) created at:
  istio.io/istio/vendor/go.opencensus.io/stats/view.init.0()
      /go/src/istio.io/istio/vendor/go.opencensus.io/stats/view/worker.go:32 +0x99
  istio.io/istio/vendor/go.opencensus.io/stats/view.init()
      <autogenerated>:1 +0xd8
  istio.io/istio/pilot/pkg/monitoring.init()
      <autogenerated>:1 +0xa1
  istio.io/istio/pilot/pkg/model.init()
      <autogenerated>:1 +0x10f
  istio.io/istio/pilot/pkg/networking/util.init()
      <autogenerated>:1 +0xec
  istio.io/istio/pilot/pkg/networking/core/v1alpha3/loadbalancer.init()
      <autogenerated>:1 +0xb0
  main.init()
      <autogenerated>:1 +0xa1

Goroutine 10 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:916 +0x65a
  istio.io/istio/pilot/pkg/networking/core/v1alpha3/loadbalancer.TestApplyLocalitySetting.func1()
      /go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer_test.go:62 +0x57d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:865 +0x163
```

```
WARNING: DATA RACE
Read at 0x000002d65620 by goroutine 7:
  istio.io/istio/vendor/go.opencensus.io/stats/view.(*collector).addSample()
      /go/src/istio.io/istio/vendor/go.opencensus.io/stats/view/collector.go:36 +0x91
  istio.io/istio/vendor/go.opencensus.io/stats/view.(*viewInternal).addSample()
      /go/src/istio.io/istio/vendor/go.opencensus.io/stats/view/view.go:161 +0x15d
  istio.io/istio/vendor/go.opencensus.io/stats/view.(*recordReq).handleCommand()
      /go/src/istio.io/istio/vendor/go.opencensus.io/stats/view/worker_commands.go:166 +0x351
  istio.io/istio/vendor/go.opencensus.io/stats/view.(*worker).start()
      /go/src/istio.io/istio/vendor/go.opencensus.io/stats/view/worker.go:156 +0x2e7

Previous write at 0x000002d65620 by goroutine 12:
  istio.io/istio/vendor/github.com/lukechampine/freeze.Slice()
      /go/src/istio.io/istio/vendor/github.com/lukechampine/freeze/freeze.go:143 +0x194
  istio.io/istio/pilot/pkg/networking/core/v1alpha3/fakes.(*frozenIstioConfigStore).List()
      /go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/fakes/frozen_istio_config_store.go:32 +0xe9
  istio.io/istio/pilot/pkg/model.(*PushContext).initDestinationRules()
      /go/src/istio.io/istio/pilot/pkg/model/push_context.go:868 +0xa3
  istio.io/istio/pilot/pkg/model.(*PushContext).InitContext()
      /go/src/istio.io/istio/pilot/pkg/model/push_context.go:584 +0x15e
  istio.io/istio/pilot/pkg/networking/core/v1alpha3.newTestEnvironment()
      /go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/cluster_test.go:351 +0x254
  istio.io/istio/pilot/pkg/networking/core/v1alpha3.buildTestClustersWithProxyMetadata()
      /go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/cluster_test.go:235 +0xbd8
  istio.io/istio/pilot/pkg/networking/core/v1alpha3.TestHTTPCircuitBreakerThresholds.func1()
      /go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/cluster_test.go:175 +0x32a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:865 +0x163

Goroutine 7 (running) created at:
  istio.io/istio/vendor/go.opencensus.io/stats/view.init.0()
      /go/src/istio.io/istio/vendor/go.opencensus.io/stats/view/worker.go:32 +0x99
  istio.io/istio/vendor/go.opencensus.io/stats/view.init()
      <autogenerated>:1 +0xd8
  istio.io/istio/pilot/pkg/monitoring.init()
      <autogenerated>:1 +0xa1
  istio.io/istio/pilot/pkg/model.init()
      <autogenerated>:1 +0x10f
  istio.io/istio/pilot/pkg/networking/core/v1alpha3.init()
      <autogenerated>:1 +0xd8
  main.init()
      <autogenerated>:1 +0xa1

Goroutine 12 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:916 +0x65a
  istio.io/istio/pilot/pkg/networking/core/v1alpha3.TestHTTPCircuitBreakerThresholds()
      /go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/cluster_test.go:90 +0x326
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:865 +0x163
```

**Comment by kramerul** [comment]
It seems that the `freeze` module can't be used together with race tests. The instrumentation, added by the `-race` flag, writes on every access including reads, which is not allowed for frozen objects.

Therefore, we disabled the `freeze` for race tests.

**Comment by jkbschmid** [comment]
I consent.

**Comment by howardjohn** [comment]
Not sure I understand the race... how is the metrics library reading the slice? You just pass it the length which is a float not even a pointer?

**Comment by howardjohn** [comment]
Thanks for fixing this!

**Pull request merged**